magda-typescript-common/src/format/Measure.ts

```diff
@@ -0,0 +1,21 @@
+import {
+    SelectedFormats
+} from "./formats"
+import { Record } from "../../dist/generated/registry/api";
+
+//TODO turn this from a functional/OOP implementation to an OOP implementation (Make it a 'pumping' object)
+export abstract class Measure {
+    /**
+     * set to False if Measure isn't able to determine filetype with the data given in constructor; dcat field is null for example
+     */
+    public canDeduceFormat: boolean = false;
```

Comment:
Could this be a function instead?

The way it is now, if someone looks at this before calling `getSelectedFormats()` first, it'll be wrong.

Another possibility would be to return it as part of the output from `getSelectedFormats()`

Reply:
1. You suggested that the Measure class should just be a function. I agree with your suggestion, because Measures are meant to be immutable processors, rather than mutable objects. Right now, if I put all of the Measure functions into 1 file, 2 problems exist: 

---All private functions, that the measure function would use (like toLowerCase for the DcatMeasure) wouldn't be contained within anything, meaning it may be messy. Is this ok? 

---2ndly my SnapshotJudge Object's role is to work out which Measure is the most reliable. To work this out I'm storing the measure class inside a Snapshot interface. If I use a function, I wouldn't be able to differentiate       between different measures, because there's no instanceof for functions. I would have to use a more dodgy solution, like say store a string of the function name and use that. Is this good design practice?

2. Really, not only should the Measure be a function instead of a class, but the SnapshotJudge (Snapshot Evaluator) shouldn't be either, because they are just functional processors of information, like an AND gate is, not objects. Should I make the SnapshotJudge a function too?

Reply:
I actually meant just making `canDeduceFormat` a function!

-but-

It makes even more sense to make these immutable processors functions, that's a good idea.

In general (at least with my background, which is... java 🤦‍♂️ ) using `instanceof` to define behaviour isn't great OOP, because when you implement an interface then that implementation should be able to work everywhere the interface does. Wherever you use `instanceof`, your code has to have knowledge of the implementation rather than the interface. So I wouldn't be too attached to doing it that way :).